arangod/Cluster/ParticipantsCacheFeature.cpp

```diff
@@ -0,0 +1,251 @@
+            auto serverId = key.copyString();
+            auto isServerGood =
+                value.get(kHealthyServerKey).isEqualString(HealthyServerValue);
+            self->_isFailed[serverId] = !isServerGood;
+            LOG_DEVEL << "Setting " << serverId << " to "
+                      << self->_isFailed[serverId];
+          }
+        }
+        return true;
+      },
+      true, true));
+}
+
+ParticipantsCacheFeature::ParticipantsCacheFeature(Server& server)
+    : ArangodFeature{server, *this} {
+  setOptional(true);
+  startsAfter<SchedulerFeature>();
+  startsAfter<ClusterFeature>();
+}
+
+void ParticipantsCacheFeature::prepare() {
+  if (ServerState::instance()->isAgent()) {
+    disable();
+  } else {
+    enable();
+  }
+}
+
+void ParticipantsCacheFeature::start() {
+  LOG_DEVEL << "ParticipantsCacheFeature started";
+  auto agencyCallbackRegistry =
+      server().getEnabledFeature<ClusterFeature>().agencyCallbackRegistry();
+  if (agencyCallbackRegistry == nullptr) {
+    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_INTERNAL,
+                                   "Expected non-null AgencyCallbackRegistry "
+                                   "when starting ParticipantsCacheFeature.");
+  }
+
+  initHealthCache();
+  _cache->start(agencyCallbackRegistry);
+
+  scheduleFlush();
+  LOG_TOPIC("42af3", DEBUG, Logger::REPLICATION2)
+      << "ParticipantsCacheFeature is ready";
+}
+
+void ParticipantsCacheFeature::stop() {
+  LOG_DEVEL << "ParticipantsCacheFeature stopped";
+  try {
+    auto agencyCallbackRegistry =
+        server().getEnabledFeature<ClusterFeature>().agencyCallbackRegistry();
+    _cache->stop(agencyCallbackRegistry);
+  } catch (std::exception const& ex) {
+    LOG_TOPIC("42af2", WARN, Logger::REPLICATION2)
+        << "caught unexpected exception while unregistering agency callback in "
+           "ParticipantsCacheFeature: "
+        << ex.what();
+  }
+  if (_flushJob) {
+    _flushJob->cancel();
+  }
+}
+
+auto ParticipantsCacheFeature::status()
+    -> std::unordered_map<std::string, bool> {
+  return _cache->getIsFailed();
+}
+
+void ParticipantsCacheFeature::flush() {
+  LOG_DEVEL << "ParticipantsCacheFeature flushed";
+  AgencyCache& agencyCache =
+      server().getEnabledFeature<ClusterFeature>().agencyCache();
+  std::shared_ptr<VPackBuilder> builder;
+  std::tie(builder, std::ignore) = agencyCache.get(kSupervisionHealthPath);
+  if (auto result = builder->slice(); !result.isNone()) {
+    TRI_ASSERT(result.isObject())
+        << " expected object in agency at " << kSupervisionHealthPath
+        << " but got " << result.toString();
+    FailureMap newMap;
+    for (auto const& [key, value] : VPackObjectIterator(result)) {
+      auto serverId = key.copyString();
+      auto isServerGood =
+          value.get(kHealthyServerKey).isEqualString(HealthyServerValue);
+      newMap[serverId] = !isServerGood;
+    }
+    _cache->reset(std::move(newMap));
```

Comment:
This looks similar to the code in the agency callback. Can we make one function out of that?

Reply:
That's a good observation, thanks!